ports/unix/main: add support for MICROPYINSPECT environment variable

dlech opened the pull request:
This adds support for a MICROPYINSPECT environment variable that works
exactly like PYTHONINSPECT.

> If this is set to a non-empty string it is equivalent to specifying the -i option.
>
> This variable can also be modified by Python code using os.environ to force inspect mode on program termination.

dlech committed:
ports/unix/main: add support for MICROPYINSPECT environment variable

This adds support for a MICROPYINSPECT environment variable that works
exactly like PYTHONINSPECT.

> If this is set to a non-empty string it is equivalent to specifying the -i option.
>
> This variable can also be modified by Python code using os.environ to force inspect mode on program termination.

dlech commented:
I was going to add this to the docs as well, but I don't see anything specific to the unix port. Should we add a "quick reference" section for unix like some of the other ports?

dlech committed:
tests/cmdline/repl_inspect: new test for -i option

This adds a new test to verify that the inspect (-i) command line option works.

dlech commented:
I've also added a test for the `-i` command line option. I would like to add one for the environment variable as well, but I don't currently see a way to set an environment variable for a specific test and there is no `os.environ` in the unix port.

tve commented:
Docs would be helpful, otherwise how can anyone else find this? I've been adding an "ESP32 notes:" paragraph in other places and @jimmo is working on adding a feature to the docs style to be able to collapse such port-specific paragraphs.

dlech committed:
docs/unix: new page for UNIX port

This adds a new quickstart page for the UNIX port that documents the
command line options and environment variables.

dlech commented:
I've updated this with some docs.

tve commented on docs/unix/quickref.rst:
```diff
@@ -0,0 +1,81 @@
+
+Usage::
+
+    micropython [ -c <command> | -m <module> | <script> ] [-i ] [ -O<level> ] [ -v ] [ -X <option> ]
```
The order here is not correct, the options must precede any filename (<script>).
Here what `micropython -h` outputs for me:
```
usage: ../micropython/ports/unix/micropython [<opts>] [-X <implopt>] [-c <command>] [<filename>]
Options:
-v : verbose (trace various operations); can be multiple
-O[N] : apply bytecode optimizations of level N

Implementation specific options (-X):
  compile-only                 -- parse and compile only
  emit={bytecode,native,viper} -- set the default code emitter
  heapsize=<n>[w][K|M] -- set the heap size for the GC (default 2097152)
```
That's not correct either (-m missing) plus incorrect ordering implication.

dlech replied:
Thanks for having a look. I have fixed the order.

stinos commented:
How about also adding the `-i` information to the help text printed when running `micopython -h` (C function `usage` in main.c)? At least for me that's the very first thing I try to get to know options.

dlech committed:
ports/unix/main: add #if guard around -v option usage

This modifies the usage() function to only print the -v option help
text when MICROPY_DEBUG_PRINTERS is enabled. The -v option requires
this option to be enabled for it to have any effect.

dlech committed:
ports/unix/main: Add help text for -i option

dlech committed:
ports/unix/main: document -m option

This changes the usage text to show the -m option and also show that
running a command, module or file are mutually exclusive.

dlech commented:
I have added some additional commits to clean up the usage output a bit.

dlech committed:
ports/unix/main: Add -h option for help.

This adds a -h option to print the usage help text and adds a new,
shorter error message that is printed when invalid arguments are given.

dlech commented:
Added `-h` option from https://github.com/micropython/micropython/pull/5592#issuecomment-581022548

dlech committed:
ports/unix/modos: implement putenv and unsetenv

dlech committed:
Add test for MICROPYINSPECT environment variable

When this variable is set to non-empty string, it triggers REPL after
a command/module/file finishes running.

The python file without the file extension is because the cmdline:
parser in run-test splits on spaces, so we can't use the -c option
since `import os` can't be written without a space.

dlech commented:
Added a test for the original commit, which also required implementing uos.putenv()

dpgeorge commented:
Thanks for the efforts so far on this. +1 for the addition of `-h` and making error messages more consistent.

The main thing to discuss is the use of `getenv`/`putenv`.  For CPython it seems that the `os.environ` mapping should be used instead.  Eg if (in CPython) you do `os.putenv('mytest', '123')` then `os.getenv('mytest')` returns `None`, even though it does set the variable, which can be seen by running `os.system('env')`.  For an explanation see this SO question: https://stackoverflow.com/questions/17705419/python-os-environ-os-putenv-usr-bin-env (basically `os.environ` caches all environment values at start-up and is not refreshed).

So... this means there would be uses of `getenv`/`putenv` that behave subtly different between uPy and CPy.  I guess the standard use case for `putenv` is to modify the environment of a spawned process, eg via `os.system(...)`, in which case this different doesn't matter because you won't use `getenv` to read back such values.  And `getenv` would usually just be used to get environment variables passed in to the script.  In that case it'd suffice for uPy to just have `getenv` and `putenv` (and `unsetenv`), instead of the more complicated `os.environ` object.

dlech commented:
I was thinking along the same lines that it would be simpler to just have `putenv` and `unsetenv` instead of implementing `environ`. As far as I can tell, `putenv` and `unsetenv` as implemented here are consistent with CPython. However, the current implementation of `getenv` is already different as you have noted.

dlech committed:
tests/cpydiff: add os module environ differences

dlech commented:
Added some CPython difference docs for os.environ

![image](https://user-images.githubusercontent.com/963645/73615871-cabbad00-45d1-11ea-98e9-72dbafb514b1.png)

dlech commented:
![image](https://user-images.githubusercontent.com/963645/73616057-56ced400-45d4-11ea-97c7-b101fc691233.png)

dpgeorge commented on docs/unix/quickref.rst:
```diff
@@ -0,0 +1,93 @@
+.. option:: <script>
+
+   Runs the file ``<script>``. The file can be a plain text ``.py`` file or
+   a compiled ``.mpy`` file.
```
I'm pretty sure you can't run .mpy files this way... you can only do it using `-m`

dlech replied:
Indeed. Fixed.

dpgeorge commented on docs/unix/quickref.rst:
```diff
@@ -0,0 +1,93 @@
+.. option:: <args>
+
+    Any additional arguments after the command, module or script will be passed
+    to `sys.argv`. This is currently only implemented for the module option.
```
`sys.argv` will work also for scripts passed without `-m`, eg `micropython foo.py arg1 arg2`

dlech replied:
Fixed.

dpgeorge commented:
> I was thinking along the same lines that it would be simpler to just have `putenv` and `unsetenv` instead of implementing `environ`.

I agree, and since you added the CPy diff tests I think that's good enough.  Really good!

dpgeorge commented:
Squashed a few of the commits (ones updating missing command-line options) and merged in 122baa678733636336689feeb4e2c8b4807649db through 74106757ac4f78609f71083a6452ce5ba7edb3c2

Thank you @dlech !
